Build edxapp production AMIs earlier in pipeline.

doctoryes opened the pull request:
https://openedx.atlassian.net/browse/TE-1789

Companion PR:
https://github.com/edx-ops/gomatic-secure/pull/70

Makes changes needed to create the pipeline system which builds production AMIs for edx/edge while the stage B/M/D process is occurring, speeding up eventual production deployments.

The existing edxapp pipeline system for deployments is preserved, as it's a critical piece of infrastructure. After the new pipeline system is used and verified, the old edxapp pipeline system can be removed.

I modified the cd_edxapp.py script to be able to make build-only, M/D-only, or B/M/D pipelines, instead of breaking that up into three different scripts.

An approximation of what this pipeline system looks like can be seen on the tools GoCD instance in the `jeskew_test` pipeline group.

@edx/pipeline-team Please review.

doctoryes commented:
@edx-ops/pipeline-team I've integrated @macdiesel 's suggested changes - and I've added some other small fixes. Please review.

macdiesel commented on config.yml:
```diff
@@ -69,6 +69,123 @@ tools:
+    input_files:
+      - *tools-admin
+      - "../gomatic-secure/gocd/vars/tools/environment-deployment-play/prod-edx-edxapp-build.yml"
+      - "../gomatic-secure/gocd/vars/tools/environment-deployments/prod-edx.yml"
```
make this an anchor.

macdiesel commented on config.yml:
```diff
@@ -69,6 +69,123 @@ tools:
+    input_files:
+      - *tools-admin
+      - "../gomatic-secure/gocd/vars/tools/environment-deployment-play/prod-edx-edxapp-migrate-deploy.yml"
+      - "../gomatic-secure/gocd/vars/tools/environment-deployments/prod-edx.yml"
```
make this an alias

macdiesel commented on config.yml:
```diff
@@ -69,6 +69,123 @@ tools:
     enabled: False
```
I looked and found a duplicate input file, dbl check the input files and convert to anchors and aliases where there are duplicates please.

doctoryes replied:
I re-worked `config.yml` quite a bit to move config files used multiple times to the top of the file as anchors. I find it easier to manage.

macdiesel commented on config.yml:
```diff
@@ -77,8 +194,8 @@ tools:
+      - *deployment-edx
     enabled: True
 
   # Stage edxapp
```
Could you add comments to mark the pipelines that we are deprecating with this work?

doctoryes replied:
Will do.

doctoryes replied:
Done.

macdiesel commented on edxpipelines/pipelines/cd_edxapp.py:
```diff
@@ -233,9 +283,46 @@ def install_pipelines(save_config_locally, dry_run, variable_files, cmd_line_var
+}
+
+
+def sort_bmd(bmd_steps):
```
We should move this to utils and add a test.

macdiesel replied:
See: https://github.com/edx/edx-gomatic/commit/88e909bad510d99ebb51da62c616535749360e67

macdiesel commented on edxpipelines/pipelines/cd_edxapp.py:
```diff
@@ -233,9 +283,46 @@ def install_pipelines(save_config_locally, dry_run, variable_files, cmd_line_var
+        raise Exception('only valid stages are b, m, d and must be one of {}'.format(valid_permutations.keys()))
+
+
+def validate_pipeline_permutations(bmd_steps):
```
Should move to utils and add a test.

macdiesel commented on edxpipelines/pipelines/cd_edxapp.py:
```diff
@@ -233,9 +283,46 @@ def install_pipelines(save_config_locally, dry_run, variable_files, cmd_line_var
+        raise Exception('Only supports total Build/Migrate/Deploy, Build-only, and Migrate/Deploy-only pipelines.')
+
+
+def determine_pipeline_names(config, bmd_steps):
```
Also move to utils?

macdiesel commented on edxpipelines/pipelines/prerelease_materials.py:
```diff
@@ -0,0 +1,123 @@
+        )
+
+    # If no upstream pipelines exist, don't install them!
+    for material in config.get('upstream_pipelines', []):
```
move to utils?

macdiesel commented on edxpipelines/pipelines/prerelease_materials.py:
```diff
@@ -0,0 +1,123 @@
+    #     ignore_patterns:
+    #     - '**/*'
+
+    for material in config['materials']:
```
Move to utils?

macdiesel commented on edxpipelines/pipelines/prerelease_materials.py:
```diff
@@ -0,0 +1,123 @@
+            )
+        )
+
+    prerelease_materials_stage = pipeline.ensure_stage(constants.PRERELEASE_MATERIALS_STAGE_NAME)
```
Could we use use the armed stage for this? https://github.com/edx/edx-gomatic/blob/master/edxpipelines/patterns/stages.py#L914

Also in order for the fan in to work correctly we **should not** set_has_manual_approval().

doctoryes replied:
Indeed, it can be an armed stage - I'll do so.

doctoryes commented:
@macdiesel I've addressed all your comments above. As discussed, the plan is to squash/merge this PR - and then proceed to pulling the pattern down into re-usable code and converting the IDA pipelines as well.

feanil commented:
:+1:

doctoryes commented on config.yml:
```diff
@@ -1,27 +1,54 @@
+  - &play-programs    "../gomatic-secure/gocd/vars/tools/plays/programs.yml"
+  - &play-discovery   "../gomatic-secure/gocd/vars/tools/plays/discovery.yml"
+
+  - &prod-stage-edxapp         "../gomatic-secure/gocd/vars/tools/environment-play/prod-stage-edxapp.yml"
```
It was introduced prior to this PR - and really means `prod-and-stage`, as opposed to `loadtest`. However, @macdiesel, it looks like `prod-stage-edxapp.yml` and `loadtest-edxapp.yml` have the same contents. Can you think of a reason why we shouldn't put the contents of those files into `gocd/vars/tools/plays/edxapp.yml`?

macdiesel replied:
a mystery... I don't see prod-stage-edxapp.yml in the other PR!

feanil replied:
I see, it's just consolidated here.  Ok, I think we should fix that name but that doesn't have to be a part of this PR.  If we can actually just consolidate all three that would be ever better.
